# Dundee Housing Datasets - Overview

## Content

**[Comprehensive CSV Analysis](#csv_analysis)**
- [Explore Housing Data](#explore_csv)
- [Clean and Analyse Housing Data](#clean_analyse_csv)
- [Summary - Relationship Between Tenure Types and Projected Build Volume](#housing_summary)
 
**[Site Capacity and Area Insights](#site_capacity)**  
- [Summary - How Site Capacity Varies with Site Size and Status](#capacity_summary)

**[GeoJSON Spatial Data Exploration](#geojson_exploration)**
- [Summary - How the Complexity of Site Boundaries May Affect Spatial Analysis](#geojson_summary)

### <a id=imports></a>Imports

In [1]:
from pathlib import Path

import pandas as pd
import json

### <a id=variables></a>Variables

In [2]:
# get notebook’s root dir
root = Path().resolve()

# get the data folder relative to the root directory
data_dir = root / 'data'

## <a id=csv_analysis></a>Comprehensive CSV Analysis

### <a id=explore_csv></a>Explore Housing Data

In [3]:
# import data
housing_df = pd.read_csv(data_dir / 'housingSites.csv')

In [4]:
housing_df.shape

(183, 40)

In [5]:
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 183 entries, 0 to 182
Data columns (total 40 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   OBJECTID                        183 non-null    int64  
 1   Site reference                  183 non-null    object 
 2   LDP2 reference                  50 non-null     object 
 3   Year site added                 183 non-null    object 
 4   Site area (ha)                  183 non-null    float64
 5   Site address                    183 non-null    object 
 6   Easting                         183 non-null    int64  
 7   Northing                        183 non-null    int64  
 8   Site type                       183 non-null    object 
 9   Site status                     183 non-null    object 
 10  Owner/Developer                 183 non-null    object 
 11  Tenure type                     183 non-null    object 
 12  Planning application reference  161 

In [32]:
housing_df.describe()

,OBJECTID,Site area (ha),Easting,Northing,Site capacity,BW plots written off,No of houses,No of flats,Plots completed in survey year,Total completions,...,Year 28/29,Year 29/30,Year 30/31,Year 31/32,Year 32/33,Year 33/34,Later Years,Total Programmed,Shape__Area,Shape__Length
count,183.000000,183.000000,183.000000,183.000000,183.000000,183.000000,183.000000,183.000000,183.000000,183.000000,...,183.000000,183.000000,183.000000,183.000000,183.000000,91.000000,183.000000,168.000000,183.000000,183.000000
mean,92.087432,1.458591,340762.650273,731643.076503,29.349727,2.016393,9.972678,9.426230,2.639344,7.584699,...,2.131148,1.655738,1.355191,0.874317,0.737705,0.626374,0.109290,22.351190,14585.911488,401.366852
std,53.110842,3.785955,3318.896695,1225.978051,54.752872,16.245532,30.407928,28.228926,10.828381,28.323721,...,8.578292,7.809304,7.858611,6.792089,6.671824,5.109139,1.042541,43.176443,37859.552900,460.659554
min,1.000000,0.004055,333577.000000,729422.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,40.554260,27.979365
25%,46.500000,0.056569,338669.000000,730651.500000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,565.692955,103.668408
50%,92.000000,0.207923,340252.000000,731374.000000,9.000000,0.000000,1.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,4.500000,2079.232346,206.965263
75%,137.500000,1.013347,343219.000000,732706.500000,30.000000,0.000000,3.000000,7.500000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,26.000000,10133.470625,493.531315
max,184.000000,24.331204,348471.000000,734790.000000,341.000000,167.000000,250.000000,276.000000,111.000000,224.000000,...,60.000000,60.000000,70.000000,70.000000,76.000000,48.000000,10.000000,276.000000,243312.037400,2470.210560


In [6]:
housing_df.head(3)

,OBJECTID,Site reference,LDP2 reference,Year site added,Site area (ha),Site address,Easting,Northing,Site type,Site status,...,Year 28/29,Year 29/30,Year 30/31,Year 31/32,Year 32/33,Year 33/34,Later Years,Total Programmed,Shape__Area,Shape__Length
0,1,200321,H13,04/01/2003 00:00,1.249780,QUEEN VICTORIA WORKS,339248,730423,Constrained,Constrained,...,0,0,0,0,0,NaN,0,0.0,12497.80011,507.873255
1,2,200611,NaN,06/05/2007 00:00,1.677644,"RIVERSIDE DRIVE, FORMER HOMEBASE SITE",339577,729422,Effective,Under Construction,...,0,0,0,0,0,0.0,0,50.0,16776.44287,1056.614546
2,3,200347,NaN,12/17/2013 12:00:00 AM,1.006731,"MONIFIETH ROAD, ARMITSTEAD",347412,731148,Effective,Under Construction,...,0,0,0,0,0,0.0,0,26.0,10067.30850,413.968631


### <a id=clean_analyse_csv></a>Clean and Analyse Housing Data

In [7]:
# function to remove extra spaces and apply capitalisation
def clean_text(text):
    if isinstance(text, str):
        return text.strip().title()
    return text

In [8]:
# apply function to clean/capitalise Owner/Developer column
housing_df['Owner/Developer'] = housing_df['Owner/Developer'].apply(clean_text)
housing_df['Owner/Developer'].head(2)

0                 Private
1    H & H Properties Ltd
Name: Owner/Developer, dtype: object

In [9]:
# apply function to clean/capitalise Site address column
housing_df['Site address'] = housing_df['Site address'].apply(clean_text)
housing_df['Site address'].head(2)

0                     Queen Victoria Works
1    Riverside Drive, Former Homebase Site
Name: Site address, dtype: object

In [10]:
# identify unique tenure types
housing_df['Tenure type'].unique()

array(['Private', 'RSL', 'DCC/RSL', 'TBC', 'Private/RSL', 'Mixed',
       'Local Authority', 'Housing Association'], dtype=object)

In [11]:
# group by tenure types and calculate total site capasity for each then convert to dict
tenure_capacity = housing_df.groupby('Tenure type')['Site capacity'].sum().to_dict()

In [12]:
tenure_capacity

{'DCC/RSL': 162,
 'Housing Association': 18,
 'Local Authority': 43,
 'Mixed': 343,
 'Private': 3503,
 'Private/RSL': 43,
 'RSL': 840,
 'TBC': 419}

In [13]:
housing_df.columns[housing_df.columns.str.startswith("Year")]

Index(['Year site added', 'Year 24/25', 'Year 25/26', 'Year 26/27',
       'Year 27/28', 'Year 28/29', 'Year 29/30', 'Year 30/31', 'Year 31/32',
       'Year 32/33', 'Year 33/34'],
      dtype='object')

In [14]:
active_sites_df = housing_df[housing_df['Site status'] == 'Under Construction']
print(f"Active site count: {len(active_sites_df)}")

Active site count: 62


In [15]:
# calculate total projected builds for 2024-2027
year_cols = ['Year 24/25', 'Year 25/26', 'Year 26/27', 'Year 27/28']

total_units = 0
for col in year_cols:
    total_units += active_sites_df[col].sum()

print(f"The total projected units for 2024-2027 is: {total_units}.")

The total projected units for 2024-2027 is: 1412.


### <a id=housing_summary></a>Summary - Relationship Between Tenure Types and Projected Build Volume

This breif analysis explored the relationship between tenure type and projected build volume within Dundee's housing site dataset. Total site capacity was calculated for each tenure type, showing that the Private sector dominates overall capacity - contributing 3503 units, far more than any other category. Social housing providers - RSLs (Registered Social Landlords) also contribute a significant volume (840 units), while Local Authority and Housing Association sites make up only a small portion of the total supply.

When this was compared with the total projected build volume for 2024-2027 (1412 units), this indicates that although the private sector holds most long-term capacity, it does not necessarily represent the majority of short-term delivery. Much of the private capacity appears to be related to sites at thier early planning stages or with long-term build periods. This suggests that tenure type likely influences the likelihood and timing of delivery. This has implications for Dundee's overall housing mix, future affordability, and the reliability of supply coming forward in the next few years.

## <a id=site_capacity></a>Site Capacity and Area Insights

In [16]:
large_sites_df = (housing_df[housing_df['Site area (ha)'] > 5]
    [['Site address', 'Owner/Developer', 'Site capacity', 'Site area (ha)', 'Site status']].sort_values(by="Site capacity", ascending=False))

In [17]:
large_sites_df.head()

,Site address,Owner/Developer,Site capacity,Site area (ha),Site status
16,"Western Gateway, South Gray",Springfield,341,19.341746,Under Construction
59,Central Waterfront *,Dundee City Council,276,23.969417,Allocated in LDP
52,"Dykes Of Gray, North West",Springfield,250,21.983300,Allocated in LDP
72,"Arbroath Road, Linlathen",Kirkwood Homes,250,15.897297,Under Construction
17,"Western Gateway, Swallow",Springfield,230,24.331204,Under Construction


In [18]:
avg_area_by_status = (large_sites_df.groupby('Site status')['Site area (ha)'].mean().sort_values(ascending=False))

In [19]:
avg_area_by_status.head()

Site status
Allocated in LDP      14.792294
Under Construction    13.319876
Planning Consent       7.487203
Site Complete          6.448581
Name: Site area (ha), dtype: float64

### <a id=capacity_summary></a>Summary - How Site Capacity Varies with Site Size and Status

The analysis of Dundee's housing sites shows a clear relationship between site size and development status, which in turn influences how much capacity each site is likely to deliver. The largest sites on average are those still 'Allocated in the Local Development Plan (LDP)' or 'Under Construction', with mean areas of around 14.8 ha and 13.3 ha respectively. 

Sites that have reached 'Planning Consent' are noticeably smaller, averaging 7.5 ha, suggesting these are mid-scale developments progressing through the planning stages. In comparison, 'Completed' sites have the smallest average area at around 6.4 ha, indicating that smaller sites tend to reach completion more quickly and contribute to short-term delivery.

Overall, larger sites mainly contribute to future potential capacity but are more likely to be in early stages of development, while smaller sites are more likely to be built already. This analysis highlights how site size influences delivery speed, with small sites providing immediate capacity, and large sites contributing to long-term supply.

## <a id=geojson_exploration></a>GeoJSON Spatial Data Exploration

In [20]:
with open(data_dir / 'housingSites.geojson') as f:
    geo_data  =json.load(f)
    
features = geo_data['features']
print(f"Total features: {len(features)}") 

Total features: 183


In [21]:
def calculate_coord_points(geo_data, feature_index):
    """Extract polygon coords of the specified feature and calculate the number of coordinate points."""
    feature = geo_data['features'][feature_index]
    geometry = feature['geometry']
    coords = geometry['coordinates']

    # if polygon type
    if geometry['type'] == 'Polygon':
        points = len(coords[0])

    # if multipolygon type (take outer ring of first polygon)  
    elif geometry['type'] == 'MultiPolygon':
        # points in the outer ring of each polygon
        points = sum(len(poly[0]) for poly in coords)

    else:
        raise ValueError(f"Unsupported geometry type: {geometry['type']}")

    print(f"There are {points} total points in feature {feature_index}.")

In [22]:
def print_coordinate_counts(geo_data, feature_index, points: int):
    """Extract polygon coords of the specified feature and print out specified number of coordinate points."""
    feature = geo_data['features'][feature_index]
    geometry = feature['geometry']
    coords = geometry['coordinates']

    # if polygon type
    if geometry['type'] == 'Polygon':
        outer_ring = coords[0]

    # if multipolygon type (take outer ring of first polygon)
    elif geometry['type'] == 'MultiPolygon':
        outer_ring = coords[0][0]

    else:
        raise ValueError(f"Unsupported geometry type: {geometry['type']}")
        
    print(f"First {points} coordinate points from feature {feature_index}:")
    for point in outer_ring[:points]:
        print(point)

In [23]:
# summarise the coordinate counts for the first 5 features
for i in range(5):
    calculate_coord_points(geo_data, i)

There are 82 total points in feature 0.
There are 115 total points in feature 1.
There are 30 total points in feature 2.
There are 20 total points in feature 3.
There are 68 total points in feature 4.


In [24]:
# print out the first 5 coords of the first feature
print_coordinate_counts(geo_data, 0, 5)

First 5 coordinate points from feature 0:
[-2.98806788442457, 56.4615734993592]
[-2.98801886817926, 56.4617479102959]
[-2.98798774673239, 56.4618616263887]
[-2.98798163190147, 56.4619384886711]
[-2.98803490301219, 56.4619588181716]


### <a id=geojson_summary></a>Summary - How the Complexity of Site Boundaries May Affect Spatial Analysis

Spatial analysis of housing sites in Dundee relies heavily on the accuracy and structure of the polygon boundaries stored within the GeoJSON data. The complexity of these boundaries increases with the number of vertices, their shape irregularity, and the detail of the surveyed outlines, which can influence both the performance of spatial operations/analysis and the reliability of any outputs.

**Performance**:
Sites with very detailed or irregular boundaries will require more processing time during spatial operations. Each additional coordinate point increases computational effort, which means that operations on larger sites may take noticeably longer compared to smaller, more compact sites.

**Accuracy**:
Boundary complexity can distort geometric measurements. Highly detailed outlines may artificially increase the computed perimeter, and very fine warping in the boundary can slightly inflate calculated area. For Dundee's housing sites, this could affect comparisons of site size, capacity-per-hectare calculations.

**Joins**:
Irregular boundaries can lead to misclassification when determining which administrative ward, neighbourhood, or land-use zone a site falls within. This may be relevant in Dundee where boundaries for electoral wards, settlement areas, and planning zones differ in their level of detail.

**Interpretation**:
Highly detailed site polygons can reduce clarity in maps, especially when many sites are shown together (e.g. completed vs in-progress vs planning). Complex boundaries may clutter the map, making it harder to visually interpret development patterns across Dundee. Simplified versions often provide a clearer overview without losing essential spatial meaning.

### Refrences
[More than you ever wanted to know about GeoJSON](https://macwright.com/2015/03/23/geojson-second-bite)<br>
[Optimising MapLibre Performance: Tips for Large GeoJSON Datasets](https://maplibre.org/maplibre-gl-js/docs/guides/large-data/)